runtime/runtime_storage.go

```diff
@@ -156,38 +171,50 @@ func (s *interpreterRuntimeStorage) writeValue(
 	key string,
 	value interpreter.OptionalValue,
 ) {
-	storageKey := storageKey{
+	fullKey := storageKey{
 		storageIdentifier: storageIdentifier,
 		key:               key,
 	}
 
 	// Only write the value to the cache.
 	// The Cache is finally written back through the runtime interface in `writeCached`
 
+	entry := s.cache[fullKey]
+	entry.mustWrite = true
+
 	switch typedValue := value.(type) {
 	case *interpreter.SomeValue:
-		s.cache[storageKey] = typedValue.Value
+		entry.value = typedValue.Value
+
 	case interpreter.NilValue:
-		s.cache[storageKey] = nil
+		entry.value = nil
+
 	default:
 		panic(errors.NewUnreachableError())
 	}
+
+	s.cache[fullKey] = entry
 }
 
 // writeCached serializes/saves all values in the cache in storage (through the runtime interface).
 //
 func (s *interpreterRuntimeStorage) writeCached() {
 
-	for storageKey, value := range s.cache {
+	for fullKey, entry := range s.cache {
+
+		if !entry.mustWrite && (entry.value == nil || !entry.value.Modified()) {
```

Comment:
If the cached value is `nil`, is it possible for there to still be a value in storage that must also be set to `nil`?

Reply:
@psiemens As far as I can see, we only have `value == nil` if the value is read and is nil; or it was written by the program and is set to nil, in which case `mustWrite` is set.

Though, I still removed ignoring writes of nil in b479c36 to be on the safe side